server/api/react.ts

```diff
@@ -0,0 +1,82 @@
+    if ("id" in req.body && "token" in req.body) {
+        // make sure we have a valid token (one that finds us a valid user)
+        if (sessions.has(req.body.token)) {
+            const username = sessions.get(req.body.token).username;
+
+            // first, lets make sure the post exists
+            db.get(`SELECT * FROM Posts WHERE ID=?`, [req.body.id], function (err: Error, postRow: any) {
+                // error running query
+                if (err) {
+                    console.log("[SQL] Error: " + err);
+                    res.status(500).send("Database error!");
+                    return;
+                }
+
+                // if no post is found
+                if (!postRow || !postRow.test_validpost) {
+                    console.log('[API] Edit request failed! (invalid post ID)');
+                    res.status(400).send("Invalid post ID!");
+                    return;
+                }
+
+                // then, check if we have already disliked this post (we want to remove the dislike if that is the case)
+                db.get(`SELECT * FROM PostDislikes WHERE ID=? AND Username=?`, [req.body.id, username], function (err: Error, dislikeRow: any) {
+                    if (err) {
+                        console.log("[SQL] Error: " + err);
+                        res.status(500).send("Database error!");
+                        return;
+                    }
+                    if (!dislikeRow || !dislikeRow.test_disliked) {
+                        // no dislike yet, we can dislike this post
+                        db.run(`INSERT INTO PostDislikes VALUES (?, ?)`, [req.body.id, username], function (err: Error) {
+                            if (err) {
+                                console.log("[SQL] Error: " + err);
+                                res.status(500).send("Database error!");
+                                return;
+                            }
+                            // successfully disliked post
+                            console.log("[API] " + username + " disliked post " + req.body.id);
+                            res.status(200).send();
+                        });
+                        return;
+                    }
+                    // post already disliked! remove dislike from the post
+                    db.run(`DELETE FROM PostDislikes WHERE ID=? AND Username=?`, [req.body.id, username], function (err: Error) {
+                        if (err) {
+                            console.log("[SQL] Error: " + err);
+                            res.status(500).send("Database error!");
+                            return;
+                        }
+                        // successfully removed dislike from post
+                        console.log("[API] " + username + " removed dislike from post " + req.body.id);
+                        res.status(200).send();
+                    });
+                    return;
```

Comment:
Instead of creating and deleting these, do an upsert ([`insert into... on conflict do update set`](https://www.sqlite.org/lang_upsert.html)) for a boolean "disliked" value. This will reduce the thing to one db.run statement.

Besides, it's a little silly to be rapidly creating and destroying objects. It's better to just set a boolean to true or false.

Reply:
Slight thing: you will also need to flip the boolean value whenever it is updated instead of inserted. You can still probably do this within the same SQL statement.